iceoryx_posh/source/popo/wait_set.cpp

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+#include "iceoryx_posh/popo/wait_set.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/condition_variable_liveliness.hpp"
+
+namespace iox
+{
+namespace popo
+{
+WaitSet::WaitSet(cxx::not_null<ConditionVariableData* const> condVarDataPtr) noexcept
+    : m_conditionVariableDataPtr(condVarDataPtr)
+    , m_conditionVariableWaiter(m_conditionVariableDataPtr)
+{
+}
+
+WaitSet::~WaitSet() noexcept
+{
+    // Notify all conditions in the vector that the condition variable data will be destroyed
+    detachAllConditions();
+    /// @todo Notify RouDi that the condition variable data shall be destroyed
+}
+
+cxx::expected<WaitSetError> WaitSet::attachCondition(Condition& condition) noexcept
+{
+    if (!condition.isConditionVariableAttached())
+    {
+        if (condition.attachConditionVariableIntern(m_conditionVariableDataPtr))
+        {
+            if (!m_conditionVector.push_back(&condition))
+            {
+                return cxx::error<WaitSetError>(WaitSetError::CONDITION_VECTOR_OVERFLOW);
+            }
+
+            ConditionVariableLiveliness condVarLiveliness{m_conditionVariableDataPtr};
+            condVarLiveliness.announce();
```

Comment:
So you only announce when there is space in the vector, But it is already set here. Feels like a problem. Do you first have to check if there is space in the vector before you attach?

Reply:
Simplified contract with the user and hence removed `ConditionVariableLiveliness `.